pkg/backup/compaction_test.go

```diff
@@ -1005,39 +939,152 @@ func getNumBackupsInRestore(t *testing.T, db *sqlutils.SQLRunner, jobID jobspb.J
 func ensureBackupExists(
-	t *testing.T, db *sqlutils.SQLRunner, collectionURIs string, start, end int64, opts string,
+	t *testing.T,
+	db *sqlutils.SQLRunner,
+	collectionURI []string,
+	start, end hlc.Timestamp,
+	opts *tree.ShowBackupOptions,
 ) {
 	t.Helper()
-	showBackupQ := fmt.Sprintf(`SHOW BACKUP FROM LATEST IN (%s)`, collectionURIs)
-	if opts != "" {
-		showBackupQ += " WITH " + opts
-	}
+	startTime, err := strconv.ParseFloat(start.AsOfSystemTime(), 64)
+	require.NoError(t, err)
+	endTime, err := strconv.ParseFloat(end.AsOfSystemTime(), 64)
+	require.NoError(t, err)
+	startNano, endNano := int64(startTime), int64(endTime)
+
 	// Convert times to millisecond epoch. We compare millisecond epoch instead of
 	// nanosecond epoch because the backup time is stored in milliseconds, but timeutil.Now()
 	// will return a nanosecond-precise epoch.
 	times := db.Query(t,
 		fmt.Sprintf(`SELECT DISTINCT
 		COALESCE(start_time::DECIMAL * 1e6, 0),
 		COALESCE(end_time::DECIMAL * 1e6, 0)
-		FROM [%s]`, showBackupQ),
-	)
+		FROM [%s]`, showBackupQuery(collectionURI, opts),
+		))
 	defer times.Close()
 	found := false
 	for times.Next() {
 		var startTime, endTime int64
 		require.NoError(t, times.Scan(&startTime, &endTime))
-		if startTime == start/1e3 && endTime == end/1e3 {
+		if startTime == startNano/1e3 && endTime == endNano/1e3 {
 			found = true
 			break
 		}
 	}
-	require.True(t, found, "missing backup with start time %d and end time %d", start, end)
+	require.True(t, found, "missing backup with start time %d and end time %d", startNano, endNano)
 }
 
-// getTime returns the current time in nanoseconds since epoch.
-func getTime(t *testing.T) int64 {
+func getTime() hlc.Timestamp {
+	return hlc.Timestamp{WallTime: timeutil.Now().UnixNano()}
+}
+
+// Some named variables to improve test readability.
+const fullCluster = ""
+
+var noAOST = hlc.Timestamp{}
+
+// fullBackupQuery creates a `BACKUP INTO` statement.
+func fullBackupQuery(
+	t *testing.T,
+	targets string,
+	collectionURI []string,
+	aost hlc.Timestamp,
+	opts *tree.BackupOptions,
+) string {
+	t.Helper()
+	uri := stringifyCollectionURI(collectionURI)
+	return fmt.Sprintf(
+		"BACKUP %s INTO (%s) %s %s",
+		targets, uri, aostExpr(t, aost), optsExpr(opts),
+	)
+}
+
+// incBackupQuery creates a `BACKUP INTO LATEST` statement against the latest
+// full backup in the collection.
+func incBackupQuery(
+	t *testing.T,
+	targets string,
+	collectionURI []string,
+	aost hlc.Timestamp,
+	opts *tree.BackupOptions,
+) string {
+	t.Helper()
+	uri := stringifyCollectionURI(collectionURI)
+	return fmt.Sprintf(
+		"BACKUP %s INTO LATEST IN (%s) %s %s",
+		targets, uri, aostExpr(t, aost), optsExpr(opts),
+	)
+}
+
+// compactBackupQuery creates a `crdb_internal.backup_compaction` statement.
+func compactionQuery(
+	t *testing.T, backupStmt string, fullPath string, start, end hlc.Timestamp,
+) string {
+	t.Helper()
+	startTime, err := strconv.ParseFloat(start.AsOfSystemTime(), 64)
```

Comment:
why are these ParseFloat calls necessary? Could you instead have:
```
return fmt.Sprintf(
		`SELECT crdb_internal.backup_compaction(
			'%s', '%s', '%s'::DECIMAL, '%s'::DECIMAL
		)`,
		backupStmt, fullPath, start.AsOfSystemTime(), end.AsOfSystemTime(),
	)
```

Similar to:https://github.com/kev-cao/cockroach/blob/backup/refactor-compaction-tests/pkg/testutils/fingerprintutils/fingerprint.go#L81

Idk if that parsing roundtrip works when the hlc contains a logical component

Reply:
Oh yeah that works. Had to update some of the logic for comparing timestamps when verifying that the compacted backup's start/end time shows up in `SHOW BACKUP` due to tomfoolery with significant figures and what not. Works on Mac, but let's see if the tests are still good on Unix since there's the whole precision difference between the two.